samples/matter/common/src/bridge/ble_connectivity_manager.cpp

```diff
@@ -247,38 +302,30 @@ CHIP_ERROR BLEConnectivityManager::Scan(ScanDoneCallback callback, void *context
 
 void BLEConnectivityManager::ReScanCallback(ScannedDevice *devices, uint8_t count, void *context)
 {
-	LOG_DBG("Lost devices no. %d", Instance().mRecovery.GetCurrentAmount());
-	LOG_DBG("Found devices no. %d", Instance().mScannedDevicesCounter);
-
-	if (Instance().mScannedDevicesCounter != 0 && !Instance().mRecovery.mRecoveryInProgress) {
-		BLEBridgedDeviceProvider *deviceLost = Instance().mRecovery.GetProvider();
-		if (deviceLost) {
-			for (uint8_t idx = 0; idx < Instance().mScannedDevicesCounter; idx++) {
-				auto &deviceScanned = Instance().mScannedDevices[idx];
-				bt_addr_le_t address = deviceLost->GetBtAddress();
-
-				if (memcmp(&deviceScanned.mAddr, &address, sizeof(address)) == 0) {
-					LOG_DBG("Found the lost device");
-
-					Instance().mRecovery.mIndexToRecover = idx;
-					Instance().mRecovery.mCurrentProvider = deviceLost;
-					Instance().mRecovery.mRecoveryInProgress = true;
-
-					DeviceLayer::PlatformMgr().ScheduleWork(
-						[](intptr_t context) {
-							Instance().Reconnect(
-								reinterpret_cast<BLEBridgedDeviceProvider *>(context));
-						},
-						reinterpret_cast<intptr_t>(deviceLost));
-					break;
-				}
-			}
-			if (!Instance().mRecovery.mRecoveryInProgress) {
-				Instance().mRecovery.NotifyProviderToRecover(deviceLost);
+	BLEBridgedDeviceProvider *provider = Instance().mRecovery.GetProvider();
+	if (provider) {
+		if (Instance().mRecovery.mAddressMatched) {
+			bt_addr_le_t *address = provider->GetBtAddress();
+
+			if (memcmp(&Instance().mRecovery.mScannedDevice.mAddr, address, sizeof(*address)) == 0) {
+				DeviceLayer::PlatformMgr().ScheduleWork(
+					[](intptr_t context) {
+						Instance().Reconnect(
+							reinterpret_cast<BLEBridgedDeviceProvider *>(context));
+					},
+					reinterpret_cast<intptr_t>(provider));
+				return;
 			}
 		}
-	} else if (Instance().mRecovery.IsNeeded()) {
-		Instance().mRecovery.StartTimer();
+		/* check if device still need to be found (for example it could be removed from the database) */
+		if (Instance().FindBLEProvider(*provider->GetBtAddress())) {
```

Comment:
I'm not sure I understand this check. Maybe there is something I'm missing. I can see it like this:
1. `BLEBridgedDeviceProvider *provider = Instance().mRecovery.GetProvider();` - get provider address
2. `if (provider) {` - it can be true (valid) or false (provider doesn't exist, e.g. it was removed)
3. Try to reconnect
4. `if (Instance().FindBLEProvider(*provider->GetBtAddress())) {` - try to get pointer to provider using a pointer to the same provider object?

Reply:
The provider got from the `Instance().mRecovery.GetProvider();` method is an instance in the ring buffer, so it possibly can not be null when we have removed the provider that existed in the bridged devices map. We should check whether the provider in the map still exists before putting it into the ring buffer again, so the 4th bullet is correct. I will add some more comments there.

Reply:
But, after rethinking, the checking of the `Instance().FindBLEProvider(provider->GetBtAddress()` should be done instead of point no. 2.